File: Scala/euler/src/re/_1r/euler/problems/102.scala
package re._1r.euler
package problems

object Euler102 {
  def mxc(a: (Double, Double), b: (Double, Double)) = {
    val m = (a._2 - b._2) / (a._1 - b._1)
    x: Double => m * x + a._2 - m * a._1
  }
  def below(l: Double => Double) = l(0) < 0
  def triangle(a: (Double, Double), b: (Double, Double), c: (Double, Double)) = {
    def within(ab: Double => Double, ac: Double => Double, bc: Double => Double) = {
      val cab = math.signum(a._1) != math.signum(b._1)
      val cac = math.signum(a._1) != math.signum(c._1)
      val cbc = math.signum(b._1) != math.signum(c._1)
      val bab = below(ab)
      val bac = below(ac)
      val bbc = below(bc)
      cab && cac && (bab ^ bac) || cab && cbc && (bab ^ bbc) || cac && cbc && (bac ^ bbc)
    }
    within(mxc(a, b), mxc(a, c), mxc(b, c))
  }
  def result = {
    val triangles = scala.io.Source.fromFile("triangles.txt").getLines.map(_.split(",")).map(_.grouped(2).map(n => (n(0).toDouble, n(1).toDouble)))
    triangles.count(tri => triangle(tri.next, tri.next, tri.next))
  }
}

File: Elixir/102.exs
defmodule Euler102 do
  def mxc0(a, b) do
    if 0 != Enum.at(a, 0) - Enum.at(b, 0) do
      Enum.at(a, 1) - ((Enum.at(a, 1) - Enum.at(b, 1)) / (Enum.at(a, 0) - Enum.at(b, 0))) * Enum.at(a, 0)
    else
      0
    end
  end
  def result do
    triangles = Enum.map(String.split(elem(File.read("triangles.txt"), 1), "\n"), fn ln -> Enum.chunk_every(Enum.map(String.split(ln, ","), fn pt -> String.to_integer(pt) end), 2) end)
    Enum.count(triangles, fn tri ->
      cab = !(Enum.at(Enum.at(tri, 0), 0) > 0 && Enum.at(Enum.at(tri, 1), 0) > 0) && !(Enum.at(Enum.at(tri, 0), 0) < 0 && Enum.at(Enum.at(tri, 1), 0) < 0)
      cac = !(Enum.at(Enum.at(tri, 0), 0) > 0 && Enum.at(Enum.at(tri, 2), 0) > 0) && !(Enum.at(Enum.at(tri, 0), 0) < 0 && Enum.at(Enum.at(tri, 2), 0) < 0)
      cbc = !(Enum.at(Enum.at(tri, 1), 0) > 0 && Enum.at(Enum.at(tri, 2), 0) > 0) && !(Enum.at(Enum.at(tri, 1), 0) < 0 && Enum.at(Enum.at(tri, 2), 0) < 0)
      bab = mxc0(Enum.at(tri, 0), Enum.at(tri, 1)) < 0
      bac = mxc0(Enum.at(tri, 0), Enum.at(tri, 2)) < 0
      bbc = mxc0(Enum.at(tri, 1), Enum.at(tri, 2)) < 0
      cab && cac && (bab && !bac || !bab && bac) || cab && cbc && (!bab && bbc || bab && !bbc) || cac && cbc && (!bac && bbc || bac && !bbc)
    end)
  end
end

#begin = System.monotonic_time(:nanosecond)
result = Euler102.result
#fin = System.monotonic_time(:nanosecond)
avg = Enum.reduce(0..500, 0, fn x, acc ->
  begin = System.monotonic_time(:nanosecond)
  Euler102.result
  fin = System.monotonic_time(:nanosecond)
  if(x >= 1) do
    (acc * (x - 1) + (fin - begin)) / x
  else
    0
  end
end) / 1000000000
IO.puts("#{result}\ntook: #{(avg)}")
